[analyzer] Detect leak of a stack address through output arguments 2/3

necto opened the pull request:
At this point, only functions called from other functions (i.e., not
top-level) are covered. Top-level functions have a different exit
sequence and will be handled by a subsequent change.

CPP-4734

-------

This is the second of three commits constituting https://github.com/llvm/llvm-project/pull/105648
it must not be merged before https://github.com/llvm/llvm-project/pull/105652

necto committed:
[analyzer] Detect leak of a stack address through output arguments

At this point only functions called from other functions (i.e., not
top-level) are covered. Top-level functions have a different exit
sequence, and will be handled by a subsequent change.

CPP-4734

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/StackAddrEscapeChecker.cpp:
```diff
@@ -297,20 +314,29 @@ std::optional<std::string> printReferrer(const MemRegion *Referrer) {
+
+  while (!Referrer->canPrintPretty()) {
+    if (const auto *SymReg = dyn_cast<SymbolicRegion>(Referrer)) {
+      Referrer = SymReg->getSymbol()->getOriginRegion()->getBaseRegion();
```
Be careful with `getOriginRegion()`, it will return null if the symbol is not a `SymbolRegionValue` or a `SymbolDerived` (e.g. a `SymbolConjured` returned by an opaque function call)!

Even when `getOriginRegion()` returns a non-null region, that region is only tangentially related to the symbol that was once stored in it. For example in a function like
```cpp
void leaker(int **arg) {
    int local;
    *arg = local;
}
void foo(int *arg) {
    int *original_arg = arg;
    arg = tweak(arg);
    leaker(&original_arg);   
}
```
the bug report would (probably) look like _"local variable 'local' is still referred to by the stack variable 'arg'"_ despite that the leaked reference is held through `original_arg` and not the new value of `arg`. Consider adding a testcase which shows this limitation.

I know that `getOriginRegion()` is currently the best tool for producing an user-readable description in situations like this, but it is fundamentally inaccurate, so I hope that eventually we could replace it with some more accurate solution (e.g. a mapping or function which maps a symbol to a variable which is currently storing that symbol).

necto replied:
> Consider adding a testcase which shows this limitation.

Added the test case: f223714

> Be careful with getOriginRegion(), it will return null if the symbol is not a SymbolRegionValue or a SymbolDerived (e.g. a SymbolConjured returned by an opaque function call)!



Added a defensive check: 4afbb63

However, I suspect it would never trigger, because I think a region containing a conjured symbol with no origin region associated would also have no known memory space so it would not reach this point because I discard such region [here](https://github.com/necto/llvm-project/blob/az/CPP-4734-stack-leak-output-arg/clang/lib/StaticAnalyzer/Checkers/StackAddrEscapeChecker.cpp#L302)

NagyDonat reviewed (changes requested):
Looks good overall, but `getOriginRegion()` is a troublesome function (see inline comment for details).

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/StackAddrEscapeChecker.cpp:
```diff
@@ -297,20 +314,29 @@ std::optional<std::string> printReferrer(const MemRegion *Referrer) {
+      Referrer = SymReg->getSymbol()->getOriginRegion()->getBaseRegion();
+    } else if (isa<CXXThisRegion>(Referrer)) {
+      // Skip members of a class, it is handled by
+      // warn_bind_ref_member_to_parameter_addr
```
I see, it seems that `CheckExprLifetime.cpp` was probably refactored, because the message that you reference does not appear and instead there are two very similarly named messages:

![image](https://github.com/user-attachments/assets/d1384f02-d375-4d5c-9469-a77b5f58bb32)

https://github.com/llvm/llvm-project/blame/f53bfa39a7dae444650a9c0e16d52301a733f5fc/clang/lib/Sema/CheckExprLifetime.cpp#L1166

You should probably update the name `warn_bind_ref_member_to_parameter_addr` to either `warn_bind_ref_member_to_parameter` or `warn_init_ptr_member_to_parameter_addr` (whichever is the right one) _and_ also mention `CheckExprLifetime.cpp` as a less accurate but more robust reference.

necto replied:
Indeed the diagnostic name is off (and in fact both of diags you mentioned fit). Corrected clarified in 47b5deecc and 4081a03fb

NagyDonat replied:
Thanks, the corrections LGTM, however they don't appear on this PR right now

necto replied:
Oups, indeed, I pushed them to the wrong branch. thank you for warning me!
they are on this branch too now: 
24bd8bffc6bb clarify reference further
9b7e3d6a80a8 Clarify reference

NagyDonat commented on clang/test/Analysis/stack-addr-ps.cpp:
```diff
@@ -161,3 +164,619 @@ C make1() {
+} // namespace leaking_as_member
+
+namespace origin_region_limitation {
+void leaker(int ***arg) {
```
Please use a different name for the parameter of `leaker` to highlight that the `arg` mentioned in the error message is the parameter of `foo()`.

necto replied:
Good point. Done 144d3e4

NagyDonat reviewed (commented):
Mostly LGTM and thanks for adding the "`origin_region_limitation`" test. (I added one minor remark as inline comment.)

NagyDonat reviewed (approved):
Thanks for the update, I'm satisfied with this commit now.

necto committed:
Merge branch 'main' into az/CPP-4734-stack-leak-output-arg

necto commented:
@steakhal I've rebased atop of `main` and squashed. CI is green. Could you, please, merge this PR (following https://github.com/llvm/llvm-project/pull/105652)?

Xazax-hun commented on clang/lib/StaticAnalyzer/Checkers/StackAddrEscapeChecker.cpp:
```diff
@@ -297,20 +314,31 @@ std::optional<std::string> printReferrer(const MemRegion *Referrer) {
+      // warn_init_ptr_member_to_parameter_addr
+      return std::nullopt;
+    } else {
+      Referrer->dump();
```
I think these dumps are not actionable to end users and we should not expose this.

necto replied:
Makes sense. Here is [the fix](https://github.com/llvm/llvm-project/pull/106232)
